Nicer compile error messages

asterite opened the pull request:
Fixes https://github.com/crystal-lang/crystal/issues/8410

This PR does two things:
1. It changes the meaning of `--error-trace`. Now the full trace is always shown, but if you pass `--error-trace` it will also show where the incorrect type came from (see https://github.com/crystal-lang/crystal/pull/11756#issuecomment-1017723360)
2. It formats error messages in a subjectively nicer way.

My end goal was to do https://github.com/crystal-lang/crystal/issues/8410#issuecomment-617417440

## What does this look like?

For this code:

```crystal
module Moo
  class Foo
    def foo(x)
      bar(x)
    end

    def bar(x)
      x << 'a'
    end
  end

  def self.moo
    Moo::Foo.new.foo(1)
  end
end

Moo.moo
```

The old error messages that showed up (using `--error-trace`) looked like this:

![image](https://user-images.githubusercontent.com/209371/150359891-d3bf7710-674d-4ef7-9ddf-d0e43586bcac.png)

Now it looks like this:

![image](https://user-images.githubusercontent.com/209371/150359979-0bec657d-2d5b-42e5-85fe-bf37b9a9643e.png)

I think overall each frame is easier to distinguish and you can see what's the typed method called together with its location (before this it was split), the colors are a bit nicer to the eye (subjective) and the "Error" text shows up only once.

## What does this look like if macros are involved?

With a code like this:

```crystal
module Moo
  class Foo
    macro some_macro
      def foo(x)
        bar(x)
      end
    end

    some_macro

    def bar(x)
      x << 'a'
    end
  end

  def self.moo
    Moo::Foo.new.foo(1)
  end
end

Moo.moo
```

Before this PR:

![image](https://user-images.githubusercontent.com/209371/150360572-6b83fd10-c07e-465c-b20d-55d97f0015dd.png)

With this PR:

![image](https://user-images.githubusercontent.com/209371/150360656-9a798c23-8c62-46eb-b585-a918743cd958.png)

I think I like it more because each frame or "concern" is clearly delimited with a blue heading.

## Why always show the full trace?

In my experience:
- In 99% of the cases without the error trace it's not clear how the error came to happen
- having to re-run the same command again with an extra argument is a bit annoying
- once I do that, I just keep `-error-trace` on because I know I'll need it
- the only problem with always showing the error trace is that there's more output. At the end of the output you will always see the same error. You can choose not to scroll up and that's it
 
## What about method/nil traces?

See https://github.com/crystal-lang/crystal/pull/11756#issuecomment-1017723360

## Are you open to suggestions about other colors/themes to use?

Of course! If you want, feel free to open a separate PR with code based on this branch, and propose anything you want. I don't want to spend too much time on this. From this code base we can start tweaking colors, spaces, words, etc., but I don't want to be in charge of doing that.

Fryguy commented:
> In 99% of the cases without the error trace it's not clear how the error came to happen
> having to re-run the same command again with an extra argument is a bit annoying
> once I do that, I just keep -error-trace on because I know I'll need it

This is exactly my workflow, and I now just have `--error-trace` all the time as well, which is annoying.  Glad to see it removed and just be on always.

My only concern here is backward compatibility...if people have automated scripts where they are passing `--error-trace`, then will this parameter cause a failure, or will it be ignored?  If the latter, then I think it's fine, and perhaps can be fully removed in 2.0.

HertzDevil commented:
Yeah the command-line option needs to stay for the rest of 1.x. However, I believe that even this new default wouldn't be enough for certain errors like "no overload matches" and `--error-trace` should probably display even more information than this new default.

asterite commented:
Ah, good point. I was thinking we can keep `--error-trace` and it would only print a message saying that this option is deprecated. At least for now.

asterite commented:
Actually I'm thinking we could leave `--error-trace`, but change its meaning to show where the incorrect type came from. That's something that's appended after the error message and if it's there all the time it might be confusing as the last thing you see isn't the place where the error happens.

Then, I'm not sure how useful is the "Nil trace" or "T trace" thing. Maybe we can just remove it.

asterite commented:
I brought back `--error-trace` but now it's used to show the type trace.

So with this file:

```crystal
module Moo
  class Foo
    def foo(x)
      a = nil
      x << qux
    end

    def qux
      a =
        if 1 == 1
          2
        else
          nil
        end
      a
    end
  end

  def self.moo
    Moo::Foo.new.foo(1)
  end
end

Moo.moo
```

Without `--error-trace` you get:

![image](https://user-images.githubusercontent.com/209371/150386685-6e8644ea-0e2c-435e-802a-c201465a0d25.png)

Then with `--error-trace` you get (just showing the bottom, the top is the same as before):

![image](https://user-images.githubusercontent.com/209371/150386787-897b4094-0401-4f41-ab7c-a3aa680af892.png)

This "Nil trace" can probably be improved, but I'd like to leave that for a separate PR (if really needed)

straight-shoota commented:
Changing what the error trace entails is orthogonal to the visual changes. The latter may be subjective, but I believe we can merge them easily. Can you please split the latter into a separate PR? 

I'm very hesitant about the changes to `--error-trace`. We must make sure to not flood the terminal with largely irrelevant error frames. That's really bad developer experience and certainly no improvement from the current behaviour.

For example, a simple type error deep in the compiler source code (see details) produces (with `--error-trace`) 50 error frames printing 500 lines of error output. But you really only need the last frame, about 10 lines, to understand the error.
<details>

```diff
--- i/src/compiler/crystal/semantic/call.cr
+++ w/src/compiler/crystal/semantic/call.cr
@@ -950,7 +950,7 @@ class Crystal::Call
               block_type = lookup_node_type(match.context, output).virtual_type
               block.type = output_type || block_type
               block.freeze_type = output_type || block_type
-              block_arg_type = program.proc_of(fun_args, block_type)
+              block_arg_type = program.proc_of(fun_args, 1)
             else
               raise "expected block to return #{output}, not #{block_type}"
             end
```

</details>

I think just making `--error-trace` the default doesn't improve anything. Instead, it even worsens the experience for long error traces.
We need more sophisticated solutions to improve this and should continue the discussion in #8410 to figure out what we can do.

asterite commented:
> For example, a simple type error deep in the compiler source code (see details) produces (with --error-trace) 50 error frames printing 500 lines of error output 

And the problem with that is...?

> Can you please split the latter into a separate PR?

No, sorry, I don't want to spend time on this if I'm not going to like the end result. I actually did the nice error messages as a bonus, what I really want is `--error-trace` on by default.

Fryguy commented:
I'm wondering if there's a middle ground here.  The specific reason that requires me to add `--error-trace` all the time is that I want to see the full stack trace.  However, I don't necessarily need a visual depiction, for every single frame, of the actual code along with which exact part of the line of code is causing the problem.  Most times I just need to know where the caller came from by file and line number.  A stack trace in Ruby is exactly that.  It's compact and it's usually good enough to give me what I need.

So, perhaps another option might be to only show the stack lines, but not the details, except for the very last frame.  For example taking @asterite 's original post what you see now is:

```
Showing last frame. Use --error-trace for full trace.

In foo.cr:8:9

 8 | x << 'a'
       ^-
Error: no overload matches 'Int32#<<' with type Char

Overloads are:
 - Int#<<(count : Int)
```

which isn't enough and is why you need `--error-trace`.  FWIW, here's what the exact same code looks like in Ruby (with literally the same code):

```
foo.cr:8:in `<<': no implicit conversion of String into Integer (TypeError)
	from foo.cr:8:in `bar'
	from foo.cr:4:in `foo'
	from foo.cr:13:in `moo'
	from foo.cr:17:in `<main>'
```

So, adding in just the stack lines, what about something like:

```
Expanding last frame only. Use --error-trace for full trace.

In foo.cr:17:5
In foo.cr:13:18
In foo.cr:4:7
In foo.cr:8:9

 8 | x << 'a'
       ^-
Error: no overload matches 'Int32#<<' with type Char

Overloads are:
 - Int#<<(count : Int)
```

then you have the full stack context.  Then I could see `--error-trace` expanding each of the frames into the visual representation.

If the file/line numbers aren't enough, I could see adding the signature somehow as well:

```
In foo.cr:17:5 'Moo:Module#moo()'
In foo.cr:13:18 'Moo::Foo#foo(Int32)'
In foo.cr:4:7 'bar(Int32)'
In foo.cr:8:9

 8 | x << 'a'
       ^-
Error: no overload matches 'Int32#<<' with type Char

Overloads are:
 - Int#<<(count : Int)
```

asterite commented:
Of course if you change one argument to not match a call's restriction, the error is obvious without `--error-trace`. The majority of the time, though, methods don't have type restrictions. Then a type slips through several layers of methods and finally fails to compile at the end. Then you have no idea where that type was introduced. Just looking at the last frame isn't helpful.

So we have four cases to consider:
1. The last frame is helpful and we don't show `--error-trace` by default: GREAT! You understand and fix the error.
2. Last frame is helpful and we show `--error-trace` by default: GREAT! You understand and fix the error (and you are left with a console filled with some more text than usual, but why would someone care about this is beyond my understanding)
3. The last frame is not helpful and we don't show `--error-trace` by default: BAD EXPERIENCE! You need to recompile your program, possibly wait for a few seconds, to understand and fix the error.
4. The last frame is not helpful and we show `--error-trace` by default: GREAT! You scroll up a bit and find the culprit line.

As you can see, `--error-trace` on by default gives a GREAT experience all the time.

The issue was that we showed "Nil trace" on by default too, so the last thing you saw on the console wasn't clear or helpful. What happened? We removed that by default, but we also removed all previous frames! Instead we should have just removed the "Nil trace."

Of course I'm just telling my story, but I'd like to understand why printing more data than "necessary" (subjective! it's always necessary for me) is a bad user experience.

asterite commented:
I've been thinking a bit more about this.

@straight-shoota You provided an example where the last frame makes sense as an error to show, and where showing all the traces isn't helpful.

Let's see some other examples:

## Array#sum

Code:

```crystal
a = ['a', 'b', 'c']
a.sum
```

Error:

```
Showing last frame. Use --error-trace for full trace.

In /usr/local/Cellar/crystal/1.3.2/src/enumerable.cr:1527:12

 1527 | type.zero
             ^---
Error: undefined method 'zero' for Char.class
```

## Array#replace

Code:

```crystal
a = ['a', 'b', 'c']
a.replace([1])
```

Error:

```
Showing last frame. Use --error-trace for full trace.

In /usr/local/Cellar/crystal/1.3.2/src/array.cr:1387:13

 1387 | @buffer.copy_from(other.to_unsafe, other.size)
                ^--------
Error: no overload matches 'Pointer(Char)#copy_from' with types Pointer(Int32), Int32

Overloads are:
 - Pointer(T)#copy_from(source : Pointer(NoReturn), count : Int)
 - Pointer(T)#copy_from(source : Pointer(T), count : Int)
```

## String#byte_at

Code:

```crystal
a = "hello"
a.byte_at('a')
```

Error:

```
Showing last frame. Use --error-trace for full trace.

In /usr/local/Cellar/crystal/1.3.2/src/string.cr:1269:32

 1269 | index += bytesize if index < 0
                                   ^
Error: no overload matches 'Char#<' with type Int32

Overloads are:
 - Char#<(other : Char)
 - Comparable(T)#<(other : T)
```

There are more examples I could show but they all look the same.

## Imagine...

Now imagine someone using the language for the first time and seeing those errors. In my mind, that would immediately drive someone away from the language.

## Thinking a bit more...

Thinking a bit more, we can see a pattern: I made a mistake in my code, but the error happened somewhere **not in my code**. In this case it's in the standard library.

So here's one idea I had: what if we show the **last frame owned by my code together with the **last frame**. The definition of "my code" would be everything in `src`, and everything in `lib` and other places is external code. I usually (never?) don't touch that, and I probably made some change to one of my files, or wrote one of my files, so it makes sense the the error shows things about my code.

I really thought the above code work, but...

Let's imagine we have a large codebase. Now the definition of "my code" is a bit more blurry. Maybe I'm writing some code that relies on classes and modules in `src` that I use but I don't know how they internally work. Now the full trace will contain frames of "my code" (the code I'm working on right now) but also code that I'm unfamiliar with, so errors in those places won't make sense to me.

## So what's the solution?

Easy! Show all frames. If the last frame isn't owned by me, I can scroll up until I find a frame owned by me. Then I can scroll between those places to better understand what's wrong.

## The problem with having to scroll

This is an exercise left for the reader.

@straight-shoota In all seriousness, I'd like to understand what's the issue with showing a lot of output, potentially having to scroll (**you will eventually have to scroll, once you recompile with --error-trace**,) and how all of that is worse than the cryptic error outputs I'm showing here.

---

@Fryguy I think showing the file and method signature in the trace is a bit more helpful than what's in `master`, but it's still not helpful when the error doesn't happen in the last frame.

oprypin commented:
:100: agree that there is no problem with scrolling. Recently I started to appreciate a particular interface that is made to be *optionally scrolled up*. Works very well.

![image](https://user-images.githubusercontent.com/371383/150651544-77f354ec-cde1-4909-8de0-fc44217f48cf.png)

And, of course, many programming languages also have large stack traces that can be optionally scrolled up.

Though I think it could be nice to try to reduce the number of lines per frame. After all, in Python that number is just 2, if I remember correctly.

straight-shoota commented:
The problem with long error traces is that is floods your shell buffer. It might even cause the buffer to overflow, then you can't even see the top of the output or go back to previous commands. Of course, this depends on your terminal configuration.
But in any case, I would not expect a compiler error to print error output with hundreds or thousands of lines. If nothing else, it makes it really hard navigating in the shell. If a program dumps so much to the console, there should be a good reason for that and the output should be relevant. I doubt there is anything relevant above a couple dozen lines of error traces. The rest is just spam. We should avoid to overload developers with lots of irrelevant output.

I'm completely aware that the current solution is bad. But I think printing the entire error trace by default is also bad. They're different kinds of bad, but I don't see either behaviour is a significant improval over the other. They both have pros and cons.
I don't want to just switch between bad states, just to get a different kind of bad.

We should strive for a good solution. I'm sure we can make this better. But it needs further analysis and exploring of options. If we think that's an important thing to do, we can make it a priority.

I had started working on improving the compiler error situation about two years ago. Most of the work so far has been internal refactoring to get to a state where we can do more powerful things with compiler errors. I have an even bigger refactoring in mind and plans for how to improve usability. But this process has been very time consuming because it touches a lot of things. Somewhere down the line I pushed it to the back because other important work promised quicker results. So it's still on the TODO list.

asterite commented:
Looking at the preliminary poll results it seems I was wrong. I'll later send a PR to just change the visuals, or maybe I'll just change this PR.

@straight-shoota what do you think about showing at most two frames by default, the last one, and the last on src if it's different that the last one? (this can be in a separate PR)

vlazar commented:
While I voted for not having `--error-trace` by default like most others on the forum, I wonder if 16 votes is representative enough to come to any conclusions. Looking to successful defaults in other languages could be more valuable, I think.

asterite commented:
Other languages don't need a trace because you have to put types in method arguments. If you have an error in a method, the error is there, it's never propagated to upper frames. This is a problem unique to Crystal. It's also the reason why I think all frames must be shown.

beta-ziliani commented:
The best way forward is to devote some time into getting good traces by default. I think the "all or nothing" will probably not get us far. We need an ML approach to get the right amount of traces 🤣 ! (Couldn't help it, sorry!)

Speaking seriously, this is an important point and since we have time to think about solution (until the 1.4 at least), it's better not to hurry and carefully consider the options, as @straight-shoota said. We put it in the priority queue to start working on it next month, although this shouldn't prevent anyone from working on it.

straight-shoota commented:
@asterite I don't think this is very unique, actually. It might be an unusual problem for a statically-typed language. But not so much for dynamically-typed languages. Apart from being a runtime error, this is very similar in Ruby, Python, JavaScript etc.

Ruby for example doesn't print hundreds of lines of error output for a nested type error. It just prints a simple stack trace. Every frame is one line with the location and label. There's not even a code extract.
I think that could work as a good basis. We can enhance it by adding some code extracts. For example, we can show that for the last frame in user code because that might be a likely origination location. Maybe we could do something fancy and highlight the locations that were changed last (file mtime or VCS dirty). If the highlighted code is in an expanded macro, we might want to show some more context, but maybe not the entire macro expansion for conciseness.
These are just a couple of idea that might help make a great DX for compiler errors. But this is hard to work out just from the drawing board. For moving forward I think we should collect actual code examples with their current error messages, and try to figure out what ideal error message we would like to get from the compiler.

asterite commented:
I understand but I think runtime errors and compile time errors are completely different things. And also when you get a nil error in Ruby you have no idea what happened. You have to do some serious debugging. With Crystal través it's very easy to figure it out, at the cost of having more information on the console.

asterite commented:
Closing until there's a consensus on the desired format.
